src/hotspot/share/utilities/istream.hpp

```diff
@@ -0,0 +1,449 @@
+// treat newline '\n' very differently from all other bytes.  Carriage
+// return '\r' is just another bit of whitespace, although it is
+// removed just before newline.
+//
+// Null '\0' is just a data byte, although it also terminates C
+// strings; the `current_line` function adds a null after removing any
+// line terminator but does not specially process any nulls embedded
+// in the line.
+//
+// There are sizing access functions which allow lines to contain
+// null, but the simpler function assumes null termination, and thus
+// lines containing null will "look" shorter when viewed as C strings.
+// Use the sizing access functions if you care about this.
+//
+// Formatting guidelines:
+//
+// Configuration data should be line-oriented.  It should be readable
+// by humans (though perhaps with difficulty).  It should be easily
+// processed by text editors and by widely available text processing
+// tools such as grep, sed, and awk.
+//
+// Configuration data should not require "compilers" to generate, if
+// possible.  It should be editable by hand, if possible.  In cases
+// where binary data is strongly required, pick a binary format
+// already native to Hotspot, such as classfile, jar, or jmod.
+//
+// Each line should be separately parseable; the parsing can be ad
+// hoc.  For constructs inherently larger than single lines (such as
+// complex method configuration information), try to use a structuring
+// principle that allows "leaf" data to be line-oriented, and delimits
+// that data with markup lines of some sort.  Try to pick a
+// line-friendly version of a standard format like XML or Markdown.
+// JSON is somewhat problematic because there is no line-friendly leaf
+// syntax: everything at the leaves must be a quoted string in JSON.
+//
+// Use simple parsing via scanf-like formats for simple applications.
+// But, keep in mind that these formats may lose data when applied to
+// unusual strings, such as class names that contain spaces, or method
+// names that contain punctuation.  For more robust transmission of
+// potentially unusual names, consider wrapping them in XML-flavored
+// lines like <tag attr='pay load'/>.
+//
+// Note: Input streams are never MT-safe.
+
+class inputStream : public CHeapObjBase {
+ public:
+  class Input;
+
+ private:
+  NONCOPYABLE(inputStream);
+
+  static constexpr size_t SMALL_SIZE =  240 DEBUG_ONLY(*0 + 10);
+  static constexpr size_t BIG_SIZE   = 2048 DEBUG_ONLY(*0 + 20);
+
+ protected:
+  // Values for _input_state, to distinguish some phases of history:
+  // Do we need to read more input (NTR)?  Did we see EOF already?
+  // Was there an error getting input or allocating buffer space?
+  enum class IState : int { NTR_STATE, EOF_STATE, ERR_STATE };
+
+  // Named offset for _next relative to _content_end, of phantom '\n'.
+  static const int NEXT_PHANTOM = 1;
+
+  Input* _input;   // where the input comes from or else nullptr
+  IState _input_state;  // one of {NTR,EOF,ERR}_STATE
```

Comment:
This comment not necessary as the type describes its valid state.

Reply:
Sometimes redundant comments are helpful.  I think this one is.  YMMV.

Reply:
The `override` keyword is nice; thank you.

I have already argued against the removal of `set_input`.  And `set_input` needs `close`.

I think `set_input` is not YAGNI but YIWNI = Yes I will need it.  The reply that “you can just wrap another i-stream around the new i-source” is fallacious because of the performance model of i-stream.

Reply:
Sorry, I'm still not on board with the `close` operation and I'm against `set_input` calling `close()` :-). Why is it necessary for the `inputStream` to require a file to be re-opened if the `inputStream` switches from one file to another?

To be clear: OK, we want `set_input` because we don't want to allocate two small buffers, that's fine by me.

Reply:
@jdksjolen I have incorporated most of your suggested changes. I left this code in for now:

```
void inputStream::set_input(inputStream::Input* input) {
  clear_buffer();
  if (_input != nullptr && _input != input) {
    _input->close();
  }
  ...
```

I am also leaning towards removing the `close()` call. Otherwise it would be unsymmetrical - the `inputStream` doesn't open the `_input` automatically, but it will close it automatically for us.

It seems better to leave both the `open` and `close` to the caller the `inputStream`.

@rose00 what do you think?